### ISSUE TITLE
Analysis reports wrong expected method call

### ISSUE THREAD

**Issue by johspaeth** [CROSSINGTUD/CryptoAnalysis#61]
On [this code](https://www.massapi.com/source/github/98/87/988707992/bundlerepository/src/main/java/org/apache/felix/bundlerepository/impl/OSGiRepositoryImpl.java.html) the analysis reports 

```
org.apache.felix.bundlerepository.impl.OSGiRepositoryImpl
     in Method: java.lang.String getSHA256(java.lang.String)
        Unexpected call to method digest. Expect a call to one of the following methods digest,update
            @r6 = virtualinvoke r2.<java.security.MessageDigest: byte[] digest()>()
```

The code calls digest() in a loop without calling reset() in between. I would expect the error to read:
`Expect a call to one of the following methods reset`

**Comment by johspaeth** [comment]
- [ ] Try to reproduce this error, either on the given code or write a similar code yourself. 
- [ ] Debug and fix it.

**Comment by Lonzak** [comment]
> Unexpected call to method <java.security.MessageDigest: byte[] digest()> on object of type java.security.MessageDigest. Expect a call to one of the following methods <java.security.MessageDigest: void update(byte[])>,<java.security.MessageDigest: void update(byte[],int,int)>,<java.security.MessageDigest: byte[] digest(byte[])>,<java.security.MessageDigest: void update(java.nio.ByteBuffer)>,<java.security.MessageDigest: void update(byte)>

I also get this error which seems to me like a False Positive?

```
MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
byte[] buf = new byte[estimatedSize];
int n;
InputStream inp = //some external stream
while ((n = inp.read(buf)) > 0) {
   messageDigest.update(buf, 0, n);
}
inp.close();
byte[] hash = messageDigest.digest(); //above errr occurs here
```

**Comment by rakshitkr** [comment]
Hello @Lonzak 
Please find an example similar to your code in the below link
https://github.com/rakshitkr/CryptoAnalysis/wiki

**Comment by johspaeth** [comment]
@enriozuni Could you please test your new rules on the example I pointed at and let me know what the output is? 
Then we can close this issue. Thanks.

**Comment by enriozuni** [comment]
@johspaeth You mean the DigestInputStream and DigestOutputStream rules? Thanks

**Comment by johspaeth** [comment]
Yes.

**Comment by enriozuni** [comment]
```java
private static String getSHA256(InputStream uri) throws IOException, NoSuchAlgorithmException
    {
        InputStream is = uri;
        MessageDigest md = MessageDigest.getInstance("SHA-256");

        // Use a digest inputstream as using byte arrays directly to compute the SHA-256 can
        // have big effects on memory consumption. I.e. you don't want to have to read the
        // entire resource in memory. We rather stream it through...
        DigestInputStream dis = new DigestInputStream(is, md);

        while (dis.read() != -1) {
            // we just drain the stream here to compute the Message Digest
        }
        
        md = dis.getMessageDigest();

        StringBuilder sb = new StringBuilder(64); // SHA-256 is always 64 hex characters
        for (byte b : md.digest())
        {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
```

I tested the code above and no errors were given from CogniCrypt analysis. The call ```dis.getMessageDigest();``` automatically does the```update()``` to the **MessageDigest** object and that is why no errors are reported. From the above example, I concluded that I need to update the **DigestInputStream** and **DigestOutputStream** rules, because both of them need the ```getMessageDigest()``` call to be mandatory, and after that doing the ```digest()``` call.

**Comment by enriozuni** [comment]
I also noticed that in the **MessageDigest** rule, in the ORDER section, there is the following statement:
```Gets, (DWOU | (Updates+, Digests)), (r, (DWOU | (Updates+, Digests)))*``` <br>
This states that the ```reset()``` must be between a ```digest()``` and an ```update()``` call. I believe that the ```reset()``` is redundant since when we do the ```digest()``` call, it does the reset of the MD object automatically. The JAVA documentation of MD, shows this as well [here](https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html).

This is the reason why there was no error given from the analysis when the code called ```digest()``` in a loop, without calling ```reset()``` in between, since the call to ```reset()``` is done automatically everytime, 
after the ```digest()``` call is made.

**Comment by johspaeth** [comment]
> I tested the code above and no errors were given from CogniCrypt analysis. The call `dis.getMessageDigest();` automatically does the`update()` to the **MessageDigest** object and that is why no errors are reported. From the above example, I concluded that I need to update the **DigestInputStream** and **DigestOutputStream** rules, because both of them need the `getMessageDigest()` call to be mandatory, and after that doing the `digest()` call.

Thanks @enriozuni then I expect you to commit your the target code (above) including a specification of how many errors you expect (probably 0) into the repository. See example [here](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/master/CryptoAnalysis/src/test/java/tests/headless/CustomRuleStopwatchTest.java). 

We speak about `reset()` on Thursday with @kruegers

**Comment by enriozuni** [comment]
> On [this code](https://www.massapi.com/source/github/98/87/988707992/bundlerepository/src/main/java/org/apache/felix/bundlerepository/impl/OSGiRepositoryImpl.java.html) the analysis reports
> 
> ```
> org.apache.felix.bundlerepository.impl.OSGiRepositoryImpl
>      in Method: java.lang.String getSHA256(java.lang.String)
>         Unexpected call to method digest. Expect a call to one of the following methods digest,update
>             @r6 = virtualinvoke r2.<java.security.MessageDigest: byte[] digest()>()
> ```
> The code calls digest() in a loop without calling reset() in between. I would expect the error to read:
> `Expect a call to one of the following methods reset`

**Small correction:**
After inspecting the code above, ```digest()``` is not called in a loop, but only one time in the end. What is in fact called in a loop is the return value of ```digest()``` call, which is of type ```byte[]```. Therefore, the analyis is correct, and does not show error markers since the ```digest()``` is only called once.
